### PR TITLE
FHB-626 UI: Build Performance Test Framework PoC

### PR DESCRIPTION
**Pull request by tgohil-hippo**
Added a Performance test PoC with tests against Find as an example.

Proof of concept framework built against Find

Choose where to store framework

Baseline tests built against find - example test

Demo'd to the dev team

### CODE REVIEW

File: src/test/k6/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+Here we have written performance tests for the three services in Family Hubs: Find, Connect, Manage.
+# Prerequisites
+
+- Install k6 locally
+- Run data seed scripts
+- Navigate to folder of service under test (i.e. find-tests, connect-tests, or manage-tests)
+
+# How to run the tests
+Set the environment variables in settings.json to define whether you are running the tests locally and which type of test you are running (load, stress, or soak)
```
Comment (Zac-AND-Digital) [inline src/test/k6/README.md:11]: I'd suggest writing some short one-liners (/ one-sentencers) on what each one of the test types means / does (load, stress, soak)

File: src/test/k6/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+k6 run main.js
+k6 run --vus=10 --duration=30s main.js
```
Comment (Zac-AND-Digital) [inline src/test/k6/README.md:13]: Likewise just a little description of what "vus" means

File: src/test/k6/tests/find-tests.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import http from 'k6/http';
+import { browser } from 'k6/browser';
+import { sleep } from 'k6';
+
+export async function searchResultsTest() {
+    const page = await browser.newPage();
+
+    try {
+        await page.goto('https://test.find-support-for-your-family.education.gov.uk/');
```
Comment (Zac-AND-Digital) [inline src/test/k6/tests/find-tests.js:9]: We can configure these tests to run locally but they are hard coded to run against the live test site here, does that option mean something different?

File: src/test/k6/tests/find-tests.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        }); //title attribute on the head rather than content body.
+        
+    } finally {
+        await page.close();
+    }
+    sleep(1);
+}
+
+export function verifyStatusCodeTest() {
+    const res = http.get('https://test.find-support-for-your-family.education.gov.uk/ServiceFilter?postcode=E1%202EN&adminarea=E09000030&latitude=51.517612&longitude=-0.056838&frompostcodesearch=True');
```
Comment (Zac-AND-Digital) [inline src/test/k6/tests/find-tests.js:33]: As above, this should probably be a config (passed in as a command line argument?), that way we can specify an environment to run against such as `local`, `test` or `test2`, and then in the code the URL can be calculated from the input.

File: src/test/k6/main.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+let TESTS_TO_RUN = [ ...TESTS ];
+
+export default function () {
+    // VU code
+    TESTS_TO_RUN.forEach(t => { t(); sleep(1); });
+    sleep(1);
+}
+
+/*
+// data teardown step
```
Comment (Zac-AND-Digital) [inline src/test/k6/main.js:47]: Very nitpicky but can be prefix this comment with TODO, e.g., `TODO: data teardown step`

**Review by Zac-Digital — APPROVED** [review]

File: src/test/k6/README.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# Web Browser Performance Testing with K6
+
+Here we have written performance tests for the three services in Family Hubs: Find, Connect, Manage.
+# Prerequisites
+
+- Install k6 locally
```
Comment (stevesatdfe) [inline src/test/k6/README.md:6]: Maybe a link/instructions to installing it could be good

File: src/test/k6/env/local/config.load.json [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+      "iterations": 400,
+      "options": {
+        "browser": {
+          "type": "chromium"
+        }
+      }
+    }
+  },
+  "thresholds": {
+    "http_req_failed": ["rate<0.01"],
```
Comment (stevesatdfe) [inline src/test/k6/env/local/config.load.json:19]: Consistency with spaces between conditions e.g. rate < 0.01

File: src/test/k6/main.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+// This function gets called automatically by K6 after a test run. Produces a test report in the location specified.
+export function handleSummary(data) {
+    console.log('Preparing the end-of-test summary...');
+    return {
+        "testResults.json": JSON.stringify(data)
+    };
+}
+
+// TODO: data teardown step
```
Comment (stevesatdfe) [inline src/test/k6/main.js:57]: Maybe have the teardown enabled but with a comment rather than a TODO commented out OR remove it completely

**Review by stevesatdfe — APPROVED** [review]

File: src/test/k6/main.js (hunk 8 above)
Comment (tgohil-hippo) [reply src/test/k6/main.js:57]: Added comments to next perf test ticket for amendments
image

**Pull request merged**